nomad/job_endpoint.go

```diff
@@ -1183,6 +1182,52 @@ func (j *Job) Scale(args *structs.JobScaleRequest, reply *structs.JobRegisterRes
 	return nil
 }
 
+func (j *Job) GetJobSubmission(args *structs.JobSubmissionRequest, reply *structs.JobSubmissionResponse) error {
+	authErr := j.srv.Authenticate(j.ctx, args)
+	if done, err := j.srv.forward("Job.GetJobSubmission", args, args, reply); done {
+		return err
+	}
+	j.srv.MeasureRPCRate("job_submission", structs.RateMetricRead, args)
+	if authErr != nil {
+		return structs.ErrPermissionDenied
+	}
+	defer metrics.MeasureSince([]string{"nomad", "job", "get_job_submission"}, time.Now())
+
+	// Check for read-job permissions
+	if aclObj, err := j.srv.ResolveACL(args); err != nil {
+		return err
+	} else if aclObj != nil && !aclObj.AllowNsOp(args.RequestNamespace(), acl.NamespaceCapabilityReadJob) {
+		return structs.ErrPermissionDenied
+	}
+
+	// Setup the blocking query
+	opts := blockingOptions{
+		queryOpts: &args.QueryOptions,
+		queryMeta: &reply.QueryMeta,
+		run: func(ws memdb.WatchSet, state *state.StateStore) error {
+			// Look for the submission
+			out, err := state.JobSubmission(ws, args.RequestNamespace(), args.JobID, args.Version)
```

Comment:
We're submitting the jobspec with the parent job and there's no pointer from `structs.Job` to the `structs.JobSubmission` (which is totally reasonable). But that means for periodic jobs when we call [`deriveJob`](https://github.com/hashicorp/nomad/blob/v1.5.3/nomad/periodic.go#L404-L432) (and similarly for dispatch jobs) we're copying the `structs.Job` and changing the job ID, which leaves no way to get from the derived job to the job submission anymore.

Maybe we should detect the case where there's a parent job ID and pass that as the job ID parameter here instead?

Reply:
I like keeping this simple and leaving that up to higher layers (UI and API). I think job invocations (aka child dispatch or periodic jobs) should be as minimal as possible, and I think reflecting that in the lowlevel RPCs/APIs is important to let users know they're minimal.

I would hate to include the parent source and give a user the impression we're "wasting" space saving the same job submission for every invocation of a job. That could lead them to disable the feature erroneously thinking they'll get a big performance improvement.

Reply:
That's reasonable, but we'll need to let the UI folks know that and we should document that you're responsible for sending the parent ID in the API docs.